experiments/speedrun/nanogpt_features_v0/README.md

```diff
@@ -0,0 +1,42 @@
+## NanoGPT_Features_v0
+
+This PR adds 3 speedruns for sizes 150m, 270m, and 460m param models with a subset of the modded-nanogpt features. The 270m mirrors the scale of NanoGPT.
+* 150m: 1.271 bpb, 2.19e17 model flops, 1.12e18 hw flops
+* 270m: 1.175 bpb, 8.7e17 model flops, 3.92e18 hw flops
+* 460m: 1.113 bpb, 2.8e18 model flops, 1.19e19 hw flops
+
+One objective here is to baseline the two repos to identify speedup opportunities. As a result, I am not ablating individual changes and instead want to add enough ML features such that the remaining speed gap can be isolated to non-ML components. For this draft, I am treating the single-file hackable transformer as an experimental messy scratchpad and leaving standardization of functions like partial rope, relu^2, and json logging integration out of scope. flops_per_token is an estimate, as lambdas are treated as rounding errors.
+
+### Features included
+*  Partial RoPE. Leave half of the head dimensions stationary. Also substantially increasing rotary frequency. 
+* QK Norm. Apply RMSNorm to Q and K.
+* Norm after embed
+* RMS Norm instead of LayerNorm
+* 2.5 TPP. Replacing the default 20x from Chinchilla (This seems drastically different? Maybe this metric needs a correction factor for embed/lm_head params?)
+* Relu^2 MLP. Acts as a computationally efficient version of ReGLU with tied weights between gate and up projection.
+* X0 Skip. 
+* exponential decay of resid. Single lambda for each layer.
+* backout lambda. Model learns single param to de-weight first 2/3 layers before lm_head projection.
+* reduced head counts. Roughly cutting head count in half and doubling head_dim
+* 0 init out projections. (May only be relevant for first 50 steps)
+* boosted attn scale. Using 1.35/sqrt(head_dim)
+
+### Some larger modeling differences with NanoGPT
+* Uses the GPT2-tokenizer with 50,000 tokens, whereas the marin-tokenizer is defaulting to 128,256 vocab-size. This means that for small models there is a substantial amount of compute locked in the lm_head projection. In terms of total param count, the 150m model has 80% of its params in the embedding and lm_head. I don't know enough about this repo yet to test other tokenizers.
```

Comment:
yeah this is a bit of a weird thing. We have evidence that the llama3 tokenzier is better than gpt2 and llama2 at even 1b scale but it does seem pretty lopsided in terms of parameter allocations at small scales